File: Android/app/src/main/java/com/dreamproject/inwords/core/RxDiffUtil.java
package com.dreamproject.inwords.core;

import android.support.v7.util.DiffUtil;
import android.util.Pair;

import java.util.Collections;
import java.util.List;

import io.reactivex.ObservableTransformer;
import io.reactivex.functions.BiFunction;

public class RxDiffUtil {
    public static <T> ObservableTransformer<List<T>, Pair<List<T>, DiffUtil.DiffResult>>
    calculate(BiFunction<List<T>, List<T>, DiffUtil.Callback> itemDiffer) {
        Pair<List<T>, DiffUtil.DiffResult> seedPair = Pair.create(Collections.emptyList(), null);
        return upstream -> upstream
                .scan(seedPair, (oldPair, nextItems) -> {
                    DiffUtil.Callback callback = itemDiffer.apply(oldPair.first, nextItems);
                    DiffUtil.DiffResult result = DiffUtil.calculateDiff(callback, true);
                    return Pair.create(nextItems, result);
                })
                .skip(1);  // downstream shouldn't receive seedPair.
    }
}

File: Android/app/src/main/java/com/dreamproject/inwords/core/RxDiffUtil.kt
package com.dreamproject.inwords.core

import android.util.Pair
import androidx.recyclerview.widget.DiffUtil
import io.reactivex.ObservableTransformer

object RxDiffUtil { //TODO remove annotations later
    fun <T> calculate(itemDiffer: (@JvmSuppressWildcards List<T>, @JvmSuppressWildcards List<T>) -> DiffUtil.Callback):
            ObservableTransformer<List<T>, Pair<List<T>, DiffUtil.DiffResult>> {
        val seedPair = Pair<List<T>, DiffUtil.DiffResult>(emptyList(), null)

        return ObservableTransformer { upstream ->
            upstream.scan(seedPair) { oldPair, nextItems ->
                val callback = itemDiffer(oldPair.first, nextItems)
                val result = DiffUtil.calculateDiff(callback, true)
                Pair(nextItems, result)
            }
                    .skip(1)
        }  // downstream shouldn't receive seedPair.
    }
}
